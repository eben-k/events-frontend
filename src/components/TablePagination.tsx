import { NavigateBefore, NavigateNext } from "@material-ui/icons";
import styled from "styled-components";
import { Typography } from "./Typography";

const PaginationContainer = styled.div`
  display: flex;
  align-items: center;

  .buttonsWrapper {
    display: flex;
    align-items: center;
    margin-left: 10px;
  }
`;

const PaginationButton = styled.button`
  height: 20px;
  width: 20px;
  margin: 5px;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  border: none;
  font-size: 14px;
  outline: none;
  background: none;

  &:first-child {
    margin-left: 0;
  }

  &:last-child {
    margin-right: 0;
  }
`;

interface PaginationProps {
  limit?: number;
  total?: number;
  page?: number;
  onPageChange: (page: number) => void;
}

const TablePagination = (props: PaginationProps) => {
  const { limit = 10, total = 0, page = 1, onPageChange } = props;

  return (
    <PaginationContainer>
      <Typography size="medium" color="darkGrey" weight="normal">
        Displaying {page * limit - limit + 1}-
        {page * limit > total ? total : page * limit} of {total}
      </Typography>
      <div className="buttonsWrapper">
        <PaginationButton
          disabled={page === 1}
          onClick={() => onPageChange(page - 1)}
        >
          <NavigateBefore />
        </PaginationButton>
        <PaginationButton
          disabled={limit * page >= total}
          onClick={() => onPageChange(page + 1)}
        >
          <NavigateNext />
        </PaginationButton>
      </div>
    </PaginationContainer>
  );
};

export default TablePagination;
