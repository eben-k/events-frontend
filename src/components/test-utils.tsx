import { FunctionComponent, ReactElement } from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore, Store } from "redux";
import { Provider } from "react-redux";
import { rootReducer, RootState } from "../store/reducers";
import AppThemeProvider from "./app/AppThemeProvider";

type WithReduxConfig = {
  initialState?: RootState;
  store?: Store<RootState>;
};

const render = (
  ui: ReactElement,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  }: WithReduxConfig = {}
) => {
  const Wrapper: FunctionComponent = ({ children }) => {
    return (
      <Provider store={store}>
        <AppThemeProvider>{children}</AppThemeProvider>
      </Provider>
    );
  };
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

export * from "@testing-library/react";

export { render };
