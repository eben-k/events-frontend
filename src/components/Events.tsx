import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import {
  fetchEvents,
  fetchEventTypes,
  fetchRecipients,
} from "../store/actions";
import EventTable from "./EventTable";
import TablePagination from "./TablePagination";
import { Typography, typographyMixin } from "./Typography";

const Container = styled.div`
  width: 100%;
  height: 100%;
  padding: 20px 0;
  overflow: auto;
  display: grid;
  grid-template-rows: auto 1fr;

  .topSection {
    padding: 0 32px;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;

    .rightFilter {
      display: grid;
      align-items: center;
      grid-auto-flow: row;
      gap: 10px;

      .select {
        width: 100%;
        display: grid;
        grid-auto-flow: column;
        gap: 5px;
        align-items: center;
      }
    }
  }
`;

const Select = styled.select`
  ${typographyMixin({ size: "small", color: "darkGrey", weight: "normal" })}
  border: none;
  padding: 10px 12px;
  background-color: ${(p) => p.theme.default.colors.accentColor};
  outline: none;
  width: 100%;
`;

const Events = () => {
  const dispatch = useDispatch();
  const totalEvents = useSelector((state) => state.events.total);
  const recipients = useSelector((state) => state.recipients.recipients);
  const eventTypes = useSelector((state) => state.eventTypes.types);
  const [page, setPage] = useState(1);
  const [recipient, setRecipient] = useState<string>();
  const [type, setType] = useState<string>();

  useEffect(() => {
    dispatch(
      fetchEvents({
        page,
        limit: 20,
        recipient_id: recipient,
        event_type: type,
      })
    );
  }, [dispatch, page, recipient, type]);

  useEffect(() => {
    dispatch(fetchRecipients());
    dispatch(fetchEventTypes());
  }, [dispatch]);

  const onSelectChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const { name, value } = e.target;
    const handler = name === "recipient" ? setRecipient : setType;

    if (value === "all") handler(undefined);
    else handler(value);
  };

  return (
    <Container>
      <div className="topSection">
        <TablePagination
          total={totalEvents}
          page={page}
          limit={20}
          onPageChange={setPage}
        />
        <div className="rightFilter">
          <div className="select">
            <Typography size="medium" color="black" weight="normal">
              Event Type:
            </Typography>
            <Select name="type" onChange={onSelectChange}>
              <option value="all">All Events</option>
              {eventTypes.map((event) => (
                <option key={event.event_type} value={event.event_type}>
                  {event.event_type}
                </option>
              ))}
            </Select>
          </div>
          <div className="select">
            <Typography size="medium" color="black" weight="normal">
              Recipients:
            </Typography>
            <Select name="recipient" onChange={onSelectChange}>
              <option value="all">All Recipients</option>
              {recipients.map((recipient) => (
                <option
                  key={recipient.recipient_id}
                  value={recipient.recipient_id}
                >
                  {recipient.recipient_id}
                </option>
              ))}
            </Select>
          </div>
        </div>
      </div>
      <EventTable />
    </Container>
  );
};

export default Events;
