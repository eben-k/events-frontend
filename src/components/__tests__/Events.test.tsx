import { render, screen } from "../test-utils";
import Events from "../Events";
import TablePagination from "../TablePagination";

describe("Events Test Suite", () => {
  it("should render table correctly", () => {
    const component = render(<Events />);
    const tableEl = component.getByRole("table");
    expect(tableEl).toMatchSnapshot();
    expect(tableEl).toBeDefined();
  });

  it("should render table pagination elements", () => {
    const component = render(
      <TablePagination total={1} page={1} limit={5} onPageChange={() => {}} />
    );
    const paginationEl = component.getAllByRole("button");

    expect(paginationEl).toBeDefined();
  });

  it("should render select elements", () => {
    const component = render(<Events />);
    const selectEl = component.getAllByRole("combobox");
    const optionEl = component.getAllByRole("option");

    expect(selectEl).toBeDefined();
    expect(optionEl).toBeDefined();
  });

  it("should get selection event types", () => {
    render(<Events />);
    expect(screen.getByText("All Events")).toBeTruthy();
  });

  it("should get selection recipients", () => {
    render(<Events />);
    expect(screen.getByText("All Recipients")).toBeTruthy();
  });
});
