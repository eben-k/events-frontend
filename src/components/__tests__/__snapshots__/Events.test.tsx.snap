// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Test Suite should render table correctly 1`] = `
<table
  aria-label="custom pagination table"
  class="sc-dkPtyc geOuGr"
>
  <thead>
    <tr
      role="row"
    >
      <td
        colspan="1"
        role="columnheader"
        style="min-width: 0;"
      >
        <p
          class="sc-bdvvaa bRgpcG"
          color="black"
        >
          ID
        </p>
      </td>
      <td
        colspan="1"
        role="columnheader"
        style="min-width: 0;"
      >
        <p
          class="sc-bdvvaa bRgpcG"
          color="black"
        >
          Care Giver
        </p>
      </td>
      <td
        colspan="1"
        role="columnheader"
        style="min-width: 0;"
      >
        <p
          class="sc-bdvvaa bRgpcG"
          color="black"
        >
          Recipient
        </p>
      </td>
      <td
        colspan="1"
        role="columnheader"
        style="min-width: 0;"
      >
        <p
          class="sc-bdvvaa bRgpcG"
          color="black"
        >
          Event
        </p>
      </td>
      <td
        colspan="1"
        role="columnheader"
        style="min-width: 0;"
      >
        <p
          class="sc-bdvvaa bRgpcG"
          color="black"
        >
          Date
        </p>
      </td>
    </tr>
  </thead>
  <tbody />
</table>
`;
